gcc/d/d-decls.cc

```diff
@@ -648,7 +648,8 @@ ClassReferenceExp::toSymbol()
       value->sym = new Symbol();
 
       // Build reference symbol.
-      tree decl = build_artificial_decl(unknown_type_node, NULL_TREE, "C");
+      tree ctype = build_ctype(type);
+      tree decl = build_artificial_decl(TREE_TYPE (ctype), NULL_TREE, "C");
```

Comment:
Because of this here, the mismatch check we have in `d_finish_symbol` triggers.

Example:
```
class A
{
    char a;
}

class B : A
{   
    int b;

    this(char a, int b) pure
    {
        this.a = a;
        this.b = b;
    }
}

const A ice = new B('E', 88);  // ICE here.
```
We are assigning a bigger value B to a smaller declaration A.

Maybe (as a distinct implementation detail) we should instead initialize these variables at runtime, rather than putting them in static data.

Not that I want to follow strictly in g++'s footsteps, but this is what they seem to be doing with types that can have a variably sized initializer.

shared: https://godbolt.org/g/FKzEay
thread-local: https://godbolt.org/g/Vl542f

We already generate module shared and static constructors, that are managed by druntime, so there's no additional scaffolding needed in place to support this.